src/core/server/saved_objects/migrationsv2/integration_tests/multiple_kibana_nodes.test.ts

```diff
@@ -0,0 +1,219 @@
+  let rootC: Root;
+
+  const migratedIndex = `.kibana_${pkg.version}_001`;
+  const fooType: SavedObjectsType = {
+    name: 'foo',
+    hidden: false,
+    mappings: { properties: { status: { type: 'text' } } },
+    namespaceType: 'agnostic',
+    migrations: {
+      '7.14.0': (doc) => {
+        if (doc.attributes?.status) {
+          doc.attributes.status = doc.attributes.status.replace('not_migrated', 'migrated');
+        }
+        return doc;
+      },
+    },
+  };
+
+  beforeAll(async () => {
+    await removeLogFiles();
+  });
+
+  afterAll(async () => {
+    await new Promise((resolve) => setTimeout(resolve, 10000));
+  });
+
+  beforeEach(async () => {
+    rootA = createRoot({
+      logFileName: Path.join(__dirname, `${LOG_FILE_PREFIX}_A.log`),
+    });
+    rootB = createRoot({
+      logFileName: Path.join(__dirname, `${LOG_FILE_PREFIX}_B.log`),
+    });
+    rootC = createRoot({
+      logFileName: Path.join(__dirname, `${LOG_FILE_PREFIX}_C.log`),
+    });
+
+    const { startES } = kbnTestServer.createTestServers({
+      adjustTimeout: (t: number) => jest.setTimeout(t),
+      settings: {
+        es: {
+          license: 'trial',
+          // original SO:
+          // [
+          //   { id: 'foo:1', type: 'foo', foo: { status: 'not_migrated_1' } },
+          //   { id: 'foo:2', type: 'foo', foo: { status: 'not_migrated_2' } },
+          //   { id: 'foo:3', type: 'foo', foo: { status: 'not_migrated_3' } },
+          // ];
+          dataArchive: Path.join(__dirname, 'archives', '7.13.0_5k_so.zip'),
+        },
+      },
+    });
+    esServer = await startES();
+  });
+
+  afterEach(async () => {
+    await Promise.all([rootA.shutdown(), rootB.shutdown(), rootC.shutdown()]);
+
+    if (esServer) {
+      await esServer.stop();
+    }
+  });
+
+  it('migrates saved objects normally when multiple Kibana instances are running', async () => {
+    const setupContracts = await Promise.all([rootA.setup(), rootB.setup(), rootC.setup()]);
+
+    setupContracts.forEach((setup) => setup.savedObjects.registerType(fooType));
+
+    const startContracts = await Promise.all([rootA.start(), rootB.start(), rootC.start()]);
+
+    const esClient = startContracts[0].elasticsearch.client.asInternalUser;
+    const migratedDocs = await fetchDocs(esClient, migratedIndex);
+
+    expect(migratedDocs.length).toBe(5000);
+    migratedDocs.forEach((doc, i) => {
+      expect(doc.id).toBe(`foo:${i}`);
+      expect(doc.foo.status).toBe(`migrated_${i}`);
+      expect(doc.migrationVersion.foo).toBe('7.14.0');
+    });
+  });
+
+  it('migrates saved objects normally when multiple Kibana instances are running with different plugins', async () => {
+    const setupContracts = await Promise.all([rootA.setup(), rootB.setup(), rootC.setup()]);
+
+    // Only register migration from one instance to simulate scenario
+    // where different plugins are enabled on different Kibana instances.
+    setupContracts[0].savedObjects.registerType(fooType);
```

Comment:
Initially I planned to create two different `Root` configurations with different plugins enabled, however this meant the best way to test was to actually perform assertions against a real migration that exists in the code today.

There is actually a limited set of plugins that both have an `enabled` config setting and recent migrations applied. Lens felt like a good candidate as they've added migrations recently, however I didn't love the idea of adding Lens sample data and testing against a real Lens migration from core.

Instead, I realized that I could simulate this more simply by only registering the SO type on one of the Kibanas. This is effectively what happens if you disable a plugin on a multi-Kibana deployment: the Kibana with the disabled plugin just never has the SO registered.

Happy to rework this if others have ideas of a better way to approach it.

Reply:
Seems like a good approach to me since it allows us to isolate this test from changes to production plugin code.